union/union.go

```diff
@@ -0,0 +1,778 @@
+	}
+}
+
+func (ud *unionDir) String() string {
+	return fmt.Sprintf("path: %s mountTable: %v children: %v", ud.path, ud.mountTable, ud.Children())
+}
+
+func (ud *unionDir) find(rel string) *baseUnionNode {
+	rel = strings.TrimPrefix(rel, "/")
+
+	if rel == "" {
+		return &ud.baseUnionNode
+	}
+
+	parts := strings.Split(rel, "/")
+	d := ud
+	var n *baseUnionNode
+
+	for _, part := range parts {
+		if d == nil {
+			return nil
+		}
+
+		children := d.Children()
+		child, ok := children[part]
+		if !ok {
+			return nil
+		}
+
+		n = child.(*baseUnionNode)
+		if cd, ok := child.(*unionDir); ok {
+			d = cd
+		}
+	}
+
+	return n
+}
+
+func (ud *unionDir) findDir(rel string) *unionDir {
+	rel = strings.TrimPrefix(rel, "/")
+
+	if rel == "" {
+		return ud
+	}
+
+	parts := strings.Split(rel, "/")
+	d := ud
+	var n *unionDir
+
+	for _, part := range parts {
+		if d == nil {
+			return nil
+		}
+
+		children := d.Children()
+		child, ok := children[part]
+		if !ok {
+			return nil
+		}
+
+		n, ok = child.(*unionDir)
+		d = n
+		if !ok {
+			return nil
+		}
+	}
+
+	return n
+}
+
+func (ud *unionDir) findFile(rel string) *unionFile {
+	rel = strings.TrimPrefix(rel, "/")
+
+	if rel == "" {
+		return nil
+	}
+
+	parts := strings.Split(rel, "/")
+	d := ud
+	var n fs.FSNode
+
+	for _, part := range parts {
+		if d == nil {
+			return nil
+		}
+
+		children := d.Children()
+		child, ok := children[part]
+		if !ok {
+			return nil
+		}
+
+		n = child
+		if cd, ok := child.(*unionDir); ok {
+			d = cd
+		}
+	}
+
+	if uf, ok := n.(*unionFile); ok {
+		return uf
+	}
+
+	return nil
+}
+
+func (ud *unionDir) RemoveFile(f fs.FSNode) error {
+	// TODO maybe someone will want to remove directories
+	uf, ok := f.(*unionFile)
+	if !ok {
+		return fmt.Errorf("cannot remove file that is not a union filesystem file")
+	}
+
+	rel, err := filepath.Rel(ud.mount.mountPoint, uf.path)
+	if err != nil {
+		return err
+	}
+
+	switch {
+	case ud.mount.c != nil:
+		return ud.mount.c.Remove(rel)
+	case ud.mount.d != nil:
+		on := ud.mount.d.find(rel)
+		if on == nil {
+			return fmt.Errorf("stale mount")
+		}
+		if on.parent == nil {
+			return fmt.Errorf("cannot remove root")
+		}
+		return on.parent.RemoveFile(on)
+	}
+
+	panic(fmt.Errorf("invalid mount table state"))
+}
+
+func (ud *unionDir) CreateFile(user, name string, perm uint32, mode uint8) (fs.File, error) {
+	// First, we find the mount that will permit creation
+	mte := ud.mount
+	if !mte.create {
+		for _, me := range ud.mountTable {
+			if me.create && ud.path == me.mountPoint {
+				mte = me
+				break
+			}
+
+			if me.replace {
+				break
+			}
+		}
+	}
+
+	if !mte.create {
+		return nil, fmt.Errorf("creation is not permitted here")
+	}
+
+	rel, err := filepath.Rel(mte.mountPoint, filepath.Join(ud.path, name))
+	if err != nil {
+		return nil, err
+	}
+
+	switch {
+	case mte.c != nil:
+		// TODO how do we pass in the user here?
+		f, err := mte.c.Create(rel, iofs.FileMode((uint32(mode)<<24)|(perm&0x00FFFFFF)))
```

Comment:
`CreateFile` on the server is used within a server to create a file for a particular user. Since the server in this case is unioning other servers over 9p, we have to use 9p semantics to create files.

`Create` in 9p has no user parameter. Whatever the logged-in user is will be the owner of the file.

Perhaps we should check if the user passed into `CreateFile` is the same as the user in the client? I'm not sure if that info is exposed through the client currently, but it would be ok to add that.

On the other hand, it's fine to union multiple file systems logged in via multiple users. Forcing the server to match the user parameter with whatever user is logged in for a particular client seems like a pain, and doesn't look like it will add any security.

I think it's fine to just ignore the user here.